### ISSUE TITLE
guidance for embedding pdb in nupkg vs snupkg

### ISSUE THREAD

**Issue by ctaggart** [dotnet/sourcelink#255]
I'd like to revisit the change in guidance from https://github.com/dotnet/sourcelink/commit/f92b277cf08caf4b86425aad1fc4cdebd63c6650 as a fix to #229.

> Including PDBs in the .nupkg is generally no longer recommended as it increases the size of the package and thus restore time for projects that consume your package, regardless of whether the user needs to debug through the source code of your library or not

This is only when publishing to NuGet.org. The only time a snupkg would be useful is when publishing to NuGet.org. If you are using any other NuGet feed such as:
- Azure Artifacts NuGet feed
- Artifactory NuGet feed
- MyGet.org feed
- AppVeyor feed
- many more...

Including PDBs in the .nupkg works all of those places and NuGet.org as well. It works at every step of development. It decreases the size of the nupkg download only slightly. The same can be said for embedded XML documentation.

## Symbol Servers need to retire
One of the most time frustrations for me is when Visual Studio debugger HTTP queries symbol servers for symbols when I want to debug something. I should be able to lock dependencies and have `dotnet restore` bring all of the available symbols asynchronously for all of the locked dependencies directly after a restore. It would be nice if all NuGet servers supported snupkg too and `dotnet restore` of symbols would look for snupkg first before querying Symbol Servers.

**Comment by tmat** [comment]
> It decreases the size of the nupkg download only slightly.  

The NuGet team collected numbers on this and the results show that this "slight" increase significantly affects download times in some regions. 

> The same can be said for embedded XML documentation.

Definitely. We could potentially use the same model for XML documentation files as for PDBs - that is move them to a "doc" server.

>  I should be able to lock dependencies and have dotnet restore bring all of the available symbols asynchronously for all of the locked dependencies directly after a restore. 

I agree. That doesn't mean Symbol Servers need to retire though. One solution is to add symbol download to `dotnet restore`, e.g.  `dotnet restore --symbols` would restore packages and download their symbols to the local symbol cache, where debugger can find them.

**Comment by TFTomSun** [comment]
@ctaggart I have exactly the case that I want to use local feeds and Artifactory instead of Nuget.org. Can you please point me to the current best approach to embed PDBs and source files? I want to use it for debugging as well as for code coverage analysis.

**Comment by gilescope** [comment]
> The NuGet team collected numbers on this and the results show that this "slight" increase significantly affects download times in some regions.

I expect that this too would be fixed in future. Meanwhile I would gladly trade slower download times for a simpler debugging experience.

**Comment by casperOne** [comment]
We are in the same situation as @TFTomSun.

We are implementing SourceLink against BitBucket, and we feel that while Artifactory may have symbol server support, it would be easier to put the PDB in the NuGet package for a few reasons:

- We don't have to worry about download times in other regions, as we're on a closed, private network.
- We don't want to force N people in the organization to update tooling to point internally to the symbol server (we do agree it can be argued that this is usually one-time-per-tool thing though).

So guidance/best practices on how to best package PDB files in NuGet assemblies would definitely be appreciated.

**Comment by tmat** [comment]
@casperOne Makes sense. It's pretty straightforward to add the PDB to the package: https://github.com/dotnet/sourcelink#alternative-pdb-distribution

**Comment by stijnherreman** [comment]
https://github.com/dotnet/sourcelink/pull/1141 has fixed this, I believe. The README now recommends symbol packages when publishing to NuGet.org, and no longer recommends against including symbols in the main package.

> This is a point in time problem. I expect that the feature will be added to all NuGet servers in future.

5 years later, [the Azure Artifacts proposal](https://developercommunity.visualstudio.com/t/add-snupkg-support-to-azure-devops-artifacts/657354) has not progressed past the initial "Under Review" stage and the various GitHub Packages feedback discussions have not received any response. According to https://github.com/NuGet/docs.microsoft.com-nuget/issues/2343#issuecomment-796313766 GitHub "have a firm stance of not supporting them", but I could not find a source for that claim.
